File: Program's_Contributed_By_Contributors/C++ Programs/MergeSort.cpp
//MERGE-SORT
#include<iostream>
using namespace std;

//merge function take two intervals 
//one from start to mid , second from mid+1 to end 
//and merge them in sorted order

void merge(int arr[], int l, int mid, int r){
    
    int n1 = mid-l+1;
    int n2 = r-mid;

    //temp arrays to store both halves of array
    int a[n1];
    int b[n2];

    for(int i=0;i<n1;i++){
        a[i]=arr[l+i];
    }
    for(int i=0;i<n2;i++){
        b[i]=arr[mid+1+i];
    }
    int i=0, j=0, k=l;

    // traverse both arrays and in each iteration add smaller of both elements in temp 
    while(i<n1 && j<n2){
        if(a[i]<b[j]){
            arr[k]=a[i];
            k++; i++;
        }
        else{
            arr[k]=b[j];
            k++; j++;
        }
    }

    //adding all left elements
    while(i<n1){
        arr[k]=a[i];
        k++; i++;
    }
    while(j<n2){
        arr[k]=b[j];
        k++; j++;
    }
}

void mergeSort(int arr[], int l, int r) {

    if(l<r) {
        int mid = (l+r) / 2;
        //dividing array into two halves
        mergeSort(arr, l, mid);
        mergeSort(arr, mid+1, r);
        merge(arr, l, mid, r);
    }
}

int main(){

    int arr[]={3,4,1,5,2};
    mergeSort(arr, 0, 4);

    for(int i=0;i<5;i++){
        cout<<arr[i]<<" ";
    }
    //1 2 3 4 5

    return 0;
}

File: Program's_Contributed_By_Contributors/Java_Programs/Data_structure/MergeSort.java
import java.util.Arrays;
import java.util.Scanner;

class MergeSort
{
    void merge(int arr[], int l, int m, int r)
    {
        int n1 = m - l + 1;
        int n2 = r - m;
        int L[] = new int [n1];
        int R[] = new int [n2];

        for (int i=0; i<n1; ++i)
            L[i] = arr[l + i];
        for (int j=0; j<n2; ++j)
            R[j] = arr[m + 1+ j];

        int i = 0, j = 0;

        int k = l;
        while (i < n1 && j < n2)
        {
            if (L[i] <= R[j])
            {
                arr[k] = L[i];
                i++;
            }
            else
            {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        while (i < n1)
        {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2)
        {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    void mergesort(int arr[], int l, int r)
    {
        if (l < r)
        {
            int m = (l+r)/2;

            mergesort(arr, l, m);
            mergesort(arr , m+1, r);

            merge(arr, l, m, r);
        }
    }

    public static void main(String args[])
    {
       Scanner sc = new Scanner(System.in);
        System.out.println("Enter size of the array :");
       int size = sc.nextInt();
       int arr[] = new int[size];
        System.out.println("Enter elements into the array: ");
        for(int i= 0;i<size;i++){
            arr[i] = sc.nextInt();
        }
        System.out.println("Array Before sorting : ");
        System.out.println(Arrays.toString(arr));

        MergeSort ob = new MergeSort();
        ob.mergesort(arr, 0, arr.length-1);

        System.out.println("Array After Sorting: ");
        System.out.println(Arrays.toString(arr));
    }
}